### PR TITLE
[WIP] core:sandbox block Module Auto-load Operations

### PR DESCRIPTION
**Pull request by tixxdz**
@poettering @kees  this is a WIP, as the related implementation may change.

Right now we only block explicit module load operations with
ProtectKernelModules=yes. This installs a seccomp on module syscalls and
removes CAP_SYS_MODULE from the capability bounding set of the unit. The
CAP_SYS_MODULE bit on its own is useless here since we already block the
related syscalls that need this capability with seccomp.

This protection only covers explicit module loading operations.
Automatic module loading due to a request for a feature that is not
loaded is not blocked. Privileged and unprivileged may trigger the kernel
to load extra modules, some modules include vulnerabilities like the
late CVE-2017-6074 DCCP double-free one.

There is a patchset under discussion that may allow to block this per
service:
http://www.openwall.com/lists/kernel-hardening/2017/04/09/1

Lets prepare for this now, the implementation may change not to be an LSM, but the exposed interface through prctl() not that much.

With this change, ProtectKernelModules now support "no|yes|full|strict"

* "no" - Classic permissions, no protection

* "yes" - Use seccomp to block module syscalls and explicit module loading

* "full" - Use 'yes' protections + if the ModAutoRestrict LSM is
           available, then module auto-loading will be allowed only if
           the following capabilities CAP_SYS_MODULE or CAP_NET_ADMIN for
           modules with a 'netdev-%s' alias are set.

* "strict" - Use 'yes' protections + removes CAP_SYS_MODULE capability +
             if ModAutoRestrict LSM is available, then module
             auto-loading will be denied.

In case the Module Auto-load Restriction is not available in the kernel,
the only noticeable change here is now CAP_SYS_MODULE capability is only
removed on "strict" instead of "yes". This should have no effect since
we already did block explicit module loading by installing seccomp
filter on module syscalls. These syscalls are the only place where
CAP_SYS_MODULE was checked, so even if CAP_SYS_MODULE is there, the
syscalls are blocked. Also in this case 'full' will have the same protection
semantics as 'yes'.

With this change systemd-networkd will continue to work and device
modules that are loaded without checking any capability will continue to
be loaded. Other services can be converted to have a "full" or "strict"
setting depending on the service.

Two services are converted here:  systemd-hostnamed and systemd-timesyncd.

The patches will be fixed later.

### CODE REVIEW

**Comment by evverx** [comment]
@tixxdz , thanks!

How can people ensure that all required modules are loaded when a service containing `ProtectKernelModules=full` is started?

**Comment by evverx** [comment]
What is the difference between adding `ProtectKernelModules=[strict|full]` to services and disabling the autoloading by doing `echo 1 > /proc/sys/kernel/modules_disabled` after all required modules have been loaded?

**Comment by kees** [comment]
Maybe another directive could be added to explicitly load modules before starting a service? (Rejected autoload modules would get logged, in theory, so admins could adjust the .service file.)

**Comment by tixxdz** [comment]
@evverx 
> How can people ensure that all required modules are loaded when a service containing ProtectKernelModules=full is started?

We have to add a logging mechanism when a module is blocked there will be an entry in logs and the service will fail. Users can check later logs, load modules manually or use systemd-modules-load.service. 

> What is the difference between adding ProtectKernelModules=[strict|full] to services and disabling the autoloading by doing echo 1 > /proc/sys/kernel/modules_disabled after all required modules have been loaded?

* per service 'full' means that you can continue to autoload modules if you have CAP_SYS_MODULE or CAP_NET_ADMIN 'netdev-%s' alias modules, maybe we may add others ... load/unload are blocked by seccomp filters.

* per service 'strict' means that you are not able to load/unload nor to autoload modules. It is disabled.

All these will not affect the rest of the system or other services. If you make sure that you have loaded all modules then you can set **modules_disabled** but you can't reset it, once set it is there until next boot. These options should be more useful. The **modules_disabled** does not only disable autoloading it disables all explicit load/unload module operations, hence if you have a system that you can't update, then you are stuck with a vulnerable module that you can't unload.

Also the use case is with **RootImage=** directive or other containers..., services will have their own root filesystem, thinking that an autoload module will load module from that root filesystem, however that's not the case, this will end up loading another module file from the host system that is on a different root filesystem. It can be that module is old or vulnerable, that's why this setting is more of control switch to say that this service should not affect global kernel state. In the other hand allowing these services or containers to autoload modules from their root filesystem will cause more harm too...

File: src/core/dbus-execute.c [hunk 1]
```diff
@@ -803,7 +804,7 @@ const sd_bus_vtable bus_exec_vtable[] = {
         SD_BUS_PROPERTY("PrivateTmp", "b", bus_property_get_bool, offsetof(ExecContext, private_tmp), SD_BUS_VTABLE_PROPERTY_CONST),
         SD_BUS_PROPERTY("PrivateDevices", "b", bus_property_get_bool, offsetof(ExecContext, private_devices), SD_BUS_VTABLE_PROPERTY_CONST),
         SD_BUS_PROPERTY("ProtectKernelTunables", "b", bus_property_get_bool, offsetof(ExecContext, protect_kernel_tunables), SD_BUS_VTABLE_PROPERTY_CONST),
-        SD_BUS_PROPERTY("ProtectKernelModules", "b", bus_property_get_bool, offsetof(ExecContext, protect_kernel_modules), SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_PROPERTY("ProtectKernelModules", "s", bus_property_get_protect_kernel_modules, offsetof(ExecContext, protect_kernel_modules), SD_BUS_VTABLE_PROPERTY_CONST),
```
Comment (poettering) [inline src/core/dbus-execute.c:807]: I fear this isn't really an OK change to make, as this breaks API to clients. The existing bool needs to stay that way, and we have to find a different way to expose this to bus clients.

Comment (poettering) [reply src/core/dbus-execute.c:807]: specifically, I figure we should add a new property "ProtectKernelModulesMode" or so, which exposes a string, and the old PKM property would expose false for the no mode, and true in all other modes.

File: src/core/protect-modules-setup.c [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+#include <linux/prctl.h>
+
+#include "protect-modules-setup.h"
+#include "string-util.h"
+#include "string-table.h"
+#include "strv.h"
+#include "util.h"
+
+int setup_module_auto_restrict(ProtectKernelModules protect_modules)
+{
```
Comment (poettering) [inline src/core/protect-modules-setup.c:35]: nitpick, our coding style says opening bracket on same line as function name

File: src/shared/lsm-util.c [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        assert(lsm_name);
+
+        r = read_one_line_file("/sys/kernel/security/lsm", &lsms);
+        if (r < 0)
+                return false;
+
+        z = strlen(lsm_name);
+        FOREACH_WORD_SEPARATOR(word, l, lsms, ",", state)
+                if (l == z && streq(word, lsm_name))
+                        return true;
```
Comment (poettering) [inline src/shared/lsm-util.c:43]: for new code we try to avoid FOREACH_WORD_SEPARATOR, and suggest people use extract_first_word() instead...

**Comment by poettering** [comment]
And also, we should only merge this when it is clear that the kernel module needed for this is going to be available in the vanilla kernels. We have been burnt too often by that...

**Comment by tixxdz** [comment]
@kees 
> Maybe another directive could be added to explicitly load modules before starting a service? (Rejected autoload modules would get logged, in theory, so admins could adjust the .service file.)

Actually you could already specify multiple `ExecStart` directives , where the first one could be prefixed by "+" 
`ExecStart=+/modprobe_path module_path`
`ExecStart=myapp_path`

This allows to run the first command with full privileges, the security sandbox mechanisms would apply only to the second directive. This also has the advantage that the first command will run in a the new mount namespace if any since we apply namespace mechanisms but not security ones, allowing to call modprobe on the appropriate module from the new mount namespace, and not the one from the other mountns.

We also have lot of other tricks, one can also use the systemd-modules-load service which comes by default:
https://www.freedesktop.org/software/systemd/man/systemd-modules-load.service.html

Use some dependencies or even use the `PartOf=` directive to be part of a target unit
https://www.freedesktop.org/software/systemd/man/systemd.unit.html#PartOf=
https://www.freedesktop.org/software/systemd/man/systemd.target.html

The `module-%s` target unit could be the synchronization point where there is a service unit to load this one and all other services that depend on it. Restarting or stopping the target unit will propagate to all others allowing to load the appropriate module and its users.

We should document this.

**Comment by tixxdz** [comment]
@poettering oups for bus API. Thanks for the review, will fix all if we manage to merge the feature.

**Comment by tixxdz** [comment]
To allow the
`ExecStart=+/modprobe_path module_path` trick

we should check https://github.com/systemd/systemd/blob/master/src/core/execute.c#L1996 on https://github.com/systemd/systemd/blob/master/src/core/execute.c#L1958 and maybe we pass `PROTECT_KERNEL_MODULES_NO` in that case, this way we don't mask `/lib/modules` here https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L107

**Comment by poettering** [comment]
@tixxdz any update on this?

**Comment by tixxdz** [comment]
@poettering @kees @evverx thanks for the review and for your precious time! unfortunately I could not follow up on this issue, and Im not up to date anymore! @poettering I won't be able to work on it.

In general, I still believe that apps or some events should not modify the privileged context of a machine, unless authorized (pre-authorization or automatic authorization) and marked. Especially in context of dynamic devices and machines that move or are shared between users / organizations =>  where you import more privileged code or bugs, I notice this in all the devices/energy sharing cases, but at the end I guess there is not time to fix all this stuff

**Comment by poettering** [comment]
OK, let's close this here then. If somebody wants to pick up the work again one day that'd be great, but for now, let's close this if there's no work going into this any time soon.

**Pull request closed without merging**